azure-iot-device/azure/iot/device/iothub/aio/async_handler_manager.py

```diff
@@ -0,0 +1,151 @@
+from concurrent.futures import ThreadPoolExecutor
+import asyncio
+import logging
+import inspect
+
+logger = logging.getLogger(__name__)
+
+MESSAGE = "_on_message_received"
+METHOD = "_on_method_request_received"
+TWIN_DP_PATCH = "_on_twin_desired_properties_patch_received"
+# TODO: add more for "event"
+
+
+class AsyncHandlerManagerException(ChainableException):
+    pass
+
+
+class AsyncHandlerManager(object):
+    def __init__(self, inbox_manager):
+        self._inbox_manager = inbox_manager
+
+        self._handler_tasks = {
+            # Inbox handler tasks
+            MESSAGE: None,
+            METHOD: None,
+            TWIN_DP_PATCH: None,
+            # Other handler tasks
+            # TODO: add
+        }
+
+        # Inbox handlers
+        self._on_message_received = None
+        self._on_method_request_received = None
+        self._on_twin_desired_properties_patch_received = None
+
+        # Other handlers
+        # TODO: add
+
+    async def _run_inbox_handler(self, inbox, handler_name):
+        """Run infinite loop that waits for an inbox to receive an object from it, then calls
+        the handler with that object
+        """
+        # Run the handler in a threadpool, so that it cannot block other handlers (from a different task),
+        # or the main client thread. The number of worker threads forms an upper bound on how many instances
+        # of the same handler can be running simultaneously.
+        tp = ThreadPoolExecutor(max_workers=4)
+        while True:
+            retval = await inbox.get()
+            # NOTE: we MUST use getattr here using the handler name, as opposed to directly passing
+            # the handler in order for the handler to be able to be updated without cancelling
+            # the running task created for this coroutine
+            handler = getattr(self, handler_name)
+            if inspect.iscoroutinefunction(handler):
+                # Wrap the coroutine in a function so it can be run in ThreadPool
+                def coro_wrapper(coro, arg):
+                    asyncio_compat.run(coro(arg))
+
+                tp.submit(coro_wrapper, handler, retval)
+            else:
+                # Run function directly in ThreadPool
+                tp.submit(handler, retval)
+
+    async def _run_event_handler(self, handler_name):
+        # TODO: implement
+        logger.error("._run_event_handler() not yet implemented")
+
+    def _get_inbox_for_handler(self, handler_name):
+        """Retrieve the inbox relevant to the handler"""
+        if handler_name == METHOD:
+            return self._inbox_manager.get_method_request_inbox()
+        elif handler_name == TWIN_DP_PATCH:
+            return self._inbox_manager.get_twin_patch_inbox()
+        elif handler_name == MESSAGE:
+            return self._inbox_manager.get_unified_message_inbox()
+        else:
+            return None
+
+    def _add_handler_task(self, handler_name):
```

Comment:
>_add_handler_task [](start = 8, length = 17)

Don't we want to store the handler _coroutine_ instead of the handler _task_?  If it's a handler _task_, I think that means it's a piece of _running_ code -- the coroutine has already been invoked..  Once the task it complete, you can't run it again.  A coroutine, on the other hand, is like a function.  you can invoke it over and over.  (disclaimer: me no good with words.  I might have this wrong.)

Reply:
No. The task is a "while True" loop that forever 1) waits for an inbox to get an item, then 2) calls the latest handler. By doing it this way it gives us a few abilities that are very useful, such as being able to update a handler without affecting the flow of the program, or be able to cancel the task if the handler is set back to None.

Because it's a while True loop, the task never completes. It runs forever until we cancel it

---
In reply to: [454574346](https://github.com/Azure/azure-iot-sdk-python/pull/620#discussion_r454574346) [](ancestors = 454574346)